File: MathCore.DSP/Filters/ChebyshevBandPass.cs
﻿using System.ComponentModel;
using System.Diagnostics;

using static System.Math;

namespace MathCore.DSP.Filters;

public class ChebyshevBandPass : ChebyshevFilter
{
    /// <summary>Формирование спецификации фильтра</summary>
    /// <param name="dt">Период дискретизации сигнала</param>
    /// <param name="fsl">Частота нижней границы полосы заграждения</param>
    /// <param name="fpl">Частота нижней границы полосы пропускания</param>
    /// <param name="fph">Частота верхней границы полосы пропускания</param>
    /// <param name="fsh">Частота верхней границы полосы заграждения</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (в единицах 0..1)</param>
    /// <param name="Gs">Неравномерность АЧХ в полосе заграждения (в единицах 0..1)</param>
    /// <returns>Спецификация фильтра</returns>
    private static Specification GetSpecification(
        double dt,
        double fsl,
        double fpl,
        double fph,
        double fsh,
        double Gp,
        double Gs)
    {
        if (Gp <= Gs)
            throw new ArgumentOutOfRangeException(
                nameof(Gp), Gp, $"Уровень АЧХ в полосе пропускания Gp={Gp} был меньше, либо равен уровню АЧХ в полосе заграждения Gs={Gs}");

        var Fsl = ToDigitalFrequency(fsl, dt);
        var Fpl = ToDigitalFrequency(fpl, dt);
        var Fph = ToDigitalFrequency(fph, dt);
        var Fsh = ToDigitalFrequency(fsh, dt);

        var Wsl = Consts.pi2 * Fsl;
        var Wpl = Consts.pi2 * Fpl;
        var Wph = Consts.pi2 * Fph;
        var Wsh = Consts.pi2 * Fsh;

        var Wc = Wpl * Wph;
        var dW = Wph - Wpl;

        // Выбор опорной частоты
        // Если   Wc / Wsh > Wsl
        // то есть      Wc > Wsl*Wsh
        // то есть Wpl*Wph > Wsl*Wsh
        // то есть центральная частота по границам подавления > центральной частоты по границам пропускания
        // то выбираем в качестве опорной частоты выбираем верхнюю границу пропускания
        // иначе, выбираем нижнюю границу пропускания
        var Ws = Wc / Wsh > Wsl
            ? Wsh
            : Wsl;
        //const double W0 = 1;                           // верхняя граница АЧХ аналогового прототипа будет всегда равна 1 рад/с
        var W1 = Abs((Wc - Ws.Pow2()) / (dW * Ws)); // пересчитываем выбранную границу в нижнюю границу пропускания АЧХ аналогового прототипа
        const double Fp = 1 / Consts.pi2;
        var Fs = W1 / Consts.pi2;

        // Для передачи информации о граничных частотах в спецификацию аналогвого прототипа перечситываем частоты цифрового фильтра обратно
        var fp = ToAnalogFrequency(Fp, dt);
        var fs = ToAnalogFrequency(Fs, dt);

        return new(dt, fp, fs, Gp, Gs);
    }

    /// <summary>Расчёт коэффициентов полиномов числителя из знаменателя передаточной функции фильтра</summary>
    /// <param name="fpl">Нижняя частота полосы пропускания</param>
    /// <param name="fph">Верхняя частота полосы пропускания</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <returns>Кортеж, содержащий массивы A - коэффициенты полинома знаменателя и B - коэффициенты полинома числителя</returns>
    private static (double[] A, double[] B) InitializeI(double fpl, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;

        var Wpl = Consts.pi2 * ToDigitalFrequency(fpl, dt);
        var Wph = Consts.pi2 * ToDigitalFrequency(fph, dt);

        var N = (int)Ceiling(arcch(Spec.kEps) / arcch(Spec.kW));
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");

        var poles = GetNormedPolesI(N, Spec.EpsP);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var pzf_poles = TransformToBandPassW(poles, Wpl, Wph);

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros = Enumerable
           .Repeat(Complex.Real, N)
           .Concat(Enumerable.Repeat(-Complex.Real, N))
           .ToArray();
        var z_poles = ToZArray(pzf_poles, dt);

        var Fpl = Wpl / Consts.pi2;
        var Fph = Wph / Consts.pi2;
        var ffp0 = ToAnalogFrequency((Fpl * Fph).Sqrt(), dt);
        var z0 = Complex.Exp(Consts.pi2 * ffp0 * dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var norm_0 = ((z0 - 1) * (z0 + 1)).Pow(N).Abs;
        var norm_p = z_poles.Multiply(z => z0 - z);

        var g_norm = N.IsEven() 
            ? Spec.Gp * (norm_p / norm_0).Abs 
            : (z0 * norm_p / norm_0).Abs;

        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = Polynom.Array.GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = Polynom.Array.GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    private static (double[] A, double[] B) InitializeII(double fpl, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;

        var Wpl = Consts.pi2 * ToDigitalFrequency(fpl, dt);
        var Wph = Consts.pi2 * ToDigitalFrequency(fph, dt);

        var N = (int)Ceiling(arcch(Spec.kEps) / arcch(Spec.kW));
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");

        var (zeros, poles) = GetNormedPolesII(N, Spec.EpsS);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var ppf_zeros_enum = TransformToBandPassW(zeros, Wpl, Wph);
        if (N.IsOdd())
            ppf_zeros_enum = ppf_zeros_enum.AppendLast(0);
        var ppf_zeros = ppf_zeros_enum.ToArray();
        var ppf_poles = TransformToBandPassW(poles, Wpl, Wph);

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros_enum = ToZ(ppf_zeros, dt);
        if (N.IsOdd())
            z_zeros_enum = z_zeros_enum.AppendLast(-1);
        var z_zeros = z_zeros_enum.ToArray();
        var z_poles = ToZArray(ppf_poles, dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var ffp0 = ToAnalogFrequency((Wpl * Wph).Sqrt() / Consts.pi2, dt);
        var z0 = Complex.Exp(Consts.pi2 * ffp0 * dt);

        var norm_0 = z_zeros.Multiply(z => z0 - z);
        var norm_p = z_poles.Multiply(z => z0 - z);

        var g_norm = N.IsEven() 
            ? Spec.Gp * (norm_p / norm_0).Re 
            : (norm_p / norm_0).Re;

        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = Polynom.Array.GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = Polynom.Array.GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    private static (double[] A, double[] B) InitializeIICorrected(double fpl, double fph, Specification Spec)
    {
        // Пересчитываем аналоговые частоты полосы заграждения в цифровые
        var dt = Spec.dt;

        var Wpl = Consts.pi2 * ToDigitalFrequency(fpl, dt);
        var Wph = Consts.pi2 * ToDigitalFrequency(fph, dt);

        var N = (int)Ceiling(arcch(Spec.kEps) / arcch(Spec.kW));
        Debug.Assert(N > 0, $"N > 0 :: {N} > 0");

        var (zeros, poles) = GetNormedPolesII(N, Spec.EpsS, Spec.kW);

        // Переносим нули и полюса аналогового нормированного ФНЧ в полосы ПЗФ
        var ppf_zeros_enum = AnalogBasedFilter.TransformToBandPassW(zeros, Wpl, Wph);
        if (N.IsOdd())
            ppf_zeros_enum = ppf_zeros_enum.AppendLast(0);
        var ppf_zeros = ppf_zeros_enum.ToArray();
        var ppf_poles = TransformToBandPassW(poles, Wpl, Wph).ToArray();

        // Преобразуем аналоговые нули и полюса в нули и полюса цифрового фильтра с помощью Z-преобразования
        var z_zeros_enum = ToZ(ppf_zeros, dt);
        if (N.IsOdd())
            z_zeros_enum = z_zeros_enum.AppendLast(-1);
        var z_zeros = z_zeros_enum.ToArray();
        var z_poles = ToZArray(ppf_poles, dt);

        // Вычисляем коэффициент нормировки фильтра на нулевой частоте 
        var ffp0 = ToAnalogFrequency((Wpl * Wph).Sqrt() / Consts.pi2, dt);
        var z0 = Complex.Exp(Consts.pi2 * ffp0 * dt);

        var norm_0 = z_zeros.Multiply(z => z0 - z);
        var norm_p = z_poles.Multiply(z => z0 - z);

        var g_norm = N.IsEven() 
            ? Spec.Gp * (norm_p / norm_0).Abs
            : (z0 * norm_p / norm_0).Abs;

        // Определяем массивы нулей коэффициентов полиномов знаменателя и числителя
        var B = Polynom.Array.GetCoefficientsInverted(z_zeros).ToArray(b => b.Re * g_norm);
        var A = Polynom.Array.GetCoefficientsInverted(z_poles).ToRe();

        return (A, B);
    }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="dt">Период дискретизации цифрового сигнала</param>
    /// <param name="fsl">Нижняя граница полосы подавления</param>
    /// <param name="fpl">Нижняя граница полосы пропускания</param>
    /// <param name="fph">Верхняя граница полосы пропускания</param>
    /// <param name="fsh">Верхняя граница полосы подавления</param>
    /// <param name="Gp">Уровень АЧХ в полосе пропускания (по умолчанию -1дБ)</param>
    /// <param name="Gs">Уровень АЧХ в полосе подавления (по умолчанию -40дБ)</param>
    /// <param name="Type">Тип фильтра I или II</param>
    public ChebyshevBandPass(
        double dt,
        double fsl,
        double fpl,
        double fph,
        double fsh,
        double Gp = 0.89125093813374556,
        double Gs = 0.01, 
        ChebyshevType Type = ChebyshevType.I)
        : this(fpl, fph, GetSpecification(dt, fsl, fpl, fph, fsh, Gp, Gs), Type) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="fpl">Нижняя граница полосы пропускания</param>
    /// <param name="fph">Верхняя граница полосы пропускания</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <param name="Type">Тип фильтра I или II</param>
    private ChebyshevBandPass(double fpl, double fph, Specification Spec, ChebyshevType Type)
        : this(Type switch
        {
            ChebyshevType.I => InitializeI(fpl, fph, Spec),
            ChebyshevType.II => InitializeII(fpl, fph, Spec),
            ChebyshevType.IICorrected => InitializeIICorrected(fpl, fph, Spec),
            _ => throw new InvalidEnumArgumentException(nameof(Type), (int)Type, typeof(ChebyshevType))
        }, Spec, Type) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="Polynoms">Кортеж с коэффициентами полиномов знаменателя и числителя функции фильтра</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <param name="Type">Тип фильтра I или II</param>
    private ChebyshevBandPass((double[] A, double[] B) Polynoms, Specification Spec, ChebyshevType Type) 
        : this(Polynoms.B, Polynoms.A, Spec, Type) { }

    /// <summary>Инициализация нового эллиптического полосозаграждающего фильтра (ПЗФ)</summary>
    /// <param name="B">Коэффициенты полинома числителя</param>
    /// <param name="A">Коэффициенты полинома знаменателя</param>
    /// <param name="Spec">Спецификация фильтра</param>
    /// <param name="Type">Тип фильтра I или II</param>
    private ChebyshevBandPass(double[] B, double[] A, Specification Spec, ChebyshevType Type) : base(B, A, Spec, Type) { }
}

File: MathCore.DSP/Filters/Python/filters/chebyshev_bandpass.py
# Полосовой фильтр Чебышева (порт C# ChebyshevBandPass)
import numpy as np
from math import ceil, pi, fabs

from .chebyshev_filter import ChebyshevFilter, ChebyshevType
from .analog_based_filter import AnalogBasedFilter, Specification
from .digital_filter import DigitalFilter

pi2 = 2.0 * pi


class ChebyshevBandPass(ChebyshevFilter):
    """
    Цифровой полосовой фильтр Чебышева (ППФ).

    Нули в z-области: N штук z=+1 (ФНЧ-прототип тип I)
    или из нулей Чебышева II рода (тип II).
    """

    @staticmethod
    def _get_specification(dt: float, fsl: float, fpl: float, fph: float, fsh: float,
                           Gp: float, Gs: float) -> Specification:
        """Спецификация ППФ через эквивалентный ФНЧ-прототип."""
        Fsl = DigitalFilter.to_digital_frequency(fsl, dt)
        Fpl = DigitalFilter.to_digital_frequency(fpl, dt)
        Fph = DigitalFilter.to_digital_frequency(fph, dt)
        Fsh = DigitalFilter.to_digital_frequency(fsh, dt)

        Wsl, Wpl = pi2 * Fsl, pi2 * Fpl
        Wph, Wsh = pi2 * Fph, pi2 * Fsh

        Wc = Wpl * Wph
        dW = Wph - Wpl

        Ws = Wsh if (Wc / Wsh > Wsl) else Wsl
        # W1 = |(Wc - Ws²) / (dW·Ws)| — прототип ФНЧ с fp=1/pi2, fs=W1/pi2
        W1 = fabs((Wc - Ws ** 2) / (dW * Ws))
        Fp = 1.0 / pi2
        Fs = W1 / pi2

        fp = DigitalFilter.to_analog_frequency(Fp, dt)
        fs = DigitalFilter.to_analog_frequency(Fs, dt)
        return Specification(dt, fp, fs, Gp, Gs)

    @staticmethod
    def _center_z(fpl: float, fph: float, dt: float) -> complex:
        """Значение z на центральной частоте полосы пропускания."""
        Fpl = DigitalFilter.to_digital_frequency(fpl, dt)
        Fph = DigitalFilter.to_digital_frequency(fph, dt)
        Fp0 = np.sqrt(Fpl * Fph)
        ffp0 = DigitalFilter.to_analog_frequency(Fp0, dt)
        return np.exp(1j * pi2 * ffp0 * dt)

    @staticmethod
    def _initialize_i(fpl: float, fph: float, spec: Specification):
        """Коэффициенты ППФ Чебышева I рода."""
        dt = spec.dt
        Wpl = pi2 * DigitalFilter.to_digital_frequency(fpl, dt)
        Wph = pi2 * DigitalFilter.to_digital_frequency(fph, dt)

        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        poles = ChebyshevFilter.get_normed_poles_i(N, spec.EpsP)

        # LP→BP трансформация полюсов
        ppf_poles = AnalogBasedFilter.transform_to_band_pass_w(poles, Wpl, Wph)

        # нули ФНЧ I рода в BP → z=+1 и z=-1 (N штук каждого)
        z_zeros = np.array([1.0 + 0j] * N + [-1.0 + 0j] * N)
        z_poles = DigitalFilter.to_z_array(ppf_poles, dt)

        # нормировка по центральной частоте
        z0 = ChebyshevBandPass._center_z(fpl, fph, dt)
        norm_0 = ((z0 - 1.0) * (z0 + 1.0)) ** N
        norm_p = np.prod([z0 - z for z in z_poles])

        k0 = spec.Gp if N % 2 == 0 else abs(z0)
        g_norm = k0 * abs(norm_p / norm_0)

        B = np.poly(z_zeros).real * g_norm
        A = np.poly(z_poles).real
        return A, B

    @staticmethod
    def _initialize_ii(fpl: float, fph: float, spec: Specification):
        """Коэффициенты ППФ Чебышева II рода."""
        dt = spec.dt
        Wpl = pi2 * DigitalFilter.to_digital_frequency(fpl, dt)
        Wph = pi2 * DigitalFilter.to_digital_frequency(fph, dt)

        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        zeros, poles = ChebyshevFilter.get_normed_poles_ii(N, spec.EpsS)

        ppf_zeros = AnalogBasedFilter.transform_to_band_pass_w(zeros, Wpl, Wph)
        if N % 2 != 0:
            ppf_zeros = list(ppf_zeros) + [0.0 + 0j]  # дополнительный нуль
        ppf_poles = AnalogBasedFilter.transform_to_band_pass_w(poles, Wpl, Wph)

        z_zeros_list = list(DigitalFilter.to_z_array(ppf_zeros, dt))
        if N % 2 != 0:
            z_zeros_list.append(-1.0 + 0j)
        z_zeros = np.array(z_zeros_list)
        z_poles = DigitalFilter.to_z_array(ppf_poles, dt)

        z0 = ChebyshevBandPass._center_z(fpl, fph, dt)
        norm_0 = np.prod([z0 - z for z in z_zeros])
        norm_p = np.prod([z0 - z for z in z_poles])
        g_norm = (spec.Gp if N % 2 == 0 else 1.0) * (norm_p / norm_0).real

        B = np.poly(z_zeros).real * g_norm
        A = np.poly(z_poles).real
        return A, B

    @staticmethod
    def _initialize_ii_corrected(fpl: float, fph: float, spec: Specification):
        """Коэффициенты ППФ Чебышева II рода с коррекцией частот."""
        dt = spec.dt
        Wpl = pi2 * DigitalFilter.to_digital_frequency(fpl, dt)
        Wph = pi2 * DigitalFilter.to_digital_frequency(fph, dt)

        N = int(ceil(ChebyshevFilter.arcch(spec.kEps) / ChebyshevFilter.arcch(spec.kW)))

        # W0 = kW для коррекции прототипа
        zeros, poles = ChebyshevFilter.get_normed_poles_ii(N, spec.EpsS, spec.kW)

        ppf_zeros = AnalogBasedFilter.transform_to_band_pass_w(zeros, Wpl, Wph)
        if N % 2 != 0:
            ppf_zeros = list(ppf_zeros) + [0.0 + 0j]
        ppf_poles = list(AnalogBasedFilter.transform_to_band_pass_w(poles, Wpl, Wph))

        z_zeros_list = list(DigitalFilter.to_z_array(ppf_zeros, dt))
        if N % 2 != 0:
            z_zeros_list.append(-1.0 + 0j)
        z_zeros = np.array(z_zeros_list)
        z_poles = DigitalFilter.to_z_array(ppf_poles, dt)

        z0 = ChebyshevBandPass._center_z(fpl, fph, dt)
        norm_0 = np.prod([z0 - z for z in z_zeros])
        norm_p = np.prod([z0 - z for z in z_poles])
        k0 = spec.Gp if N % 2 == 0 else abs(z0)
        g_norm = k0 * abs(norm_p / norm_0)

        B = np.poly(z_zeros).real * g_norm
        A = np.poly(z_poles).real
        return A, B

    def __init__(self, dt: float, fsl: float, fpl: float, fph: float, fsh: float,
                 Gp: float = 0.891250938, Gs: float = 0.01,
                 filter_type: ChebyshevType = ChebyshevType.I):
        """
        Создаёт ППФ Чебышева.

        Параметры
        ---------
        dt          : период дискретизации (с)
        fsl         : нижняя граница полосы заграждения (Гц)
        fpl         : нижняя граница полосы пропускания (Гц)
        fph         : верхняя граница полосы пропускания (Гц)
        fsh         : верхняя граница полосы заграждения (Гц)
        Gp          : коэффициент передачи на границе пропускания
        Gs          : коэффициент передачи на границе заграждения
        filter_type : тип фильтра
        """
        spec = ChebyshevBandPass._get_specification(dt, fsl, fpl, fph, fsh, Gp, Gs)
        init = {
            ChebyshevType.I:           lambda s: ChebyshevBandPass._initialize_i(fpl, fph, s),
            ChebyshevType.II:          lambda s: ChebyshevBandPass._initialize_ii(fpl, fph, s),
            ChebyshevType.IICorrected: lambda s: ChebyshevBandPass._initialize_ii_corrected(fpl, fph, s),
        }
        A, B = init[filter_type](spec)
        super().__init__(B, A, spec, filter_type)
